FUSETOOLS2-706 - adding new copyFileTextToClipboardCommand

bfitzpat opened the pull request:
PR for https://issues.redhat.com/browse/FUSETOOLS2-706
Creates a new command - vscode.didact.copyFileTextToClipboardCommand - that takes the contents of a file and places it on the clipboard
Documents the new command in the commands list and adds a new example in examples/clipboard.example.didact.md
Adds two tests, one for testing the older command (vscode.didact.copyToClipboardCommand) since one didn't already exist and one for testing the new command (vscode.didact.copyFileTextToClipboardCommand)

Signed-off-by: bfitzpat@redhat.com <bfitzpat@redhat.com>

bfitzpat commented:
Also pulled in tests from @apupier 's PR -- https://github.com/redhat-developer/vscode-didact/pull/206

apupier commented on src/extensionFunctions.ts:
```diff
@@ -896,4 +899,41 @@ export function sendTextToNamedOutputChannel(message: string, channelName?: stri
+	let testForPrefix : string = uri.toString();
+	let testUri : vscode.Uri = uri;
+	if (testForPrefix && !testForPrefix.startsWith('didact://?')) {
+		testForPrefix = 'didact://?&' + testForPrefix; // preface so we can use the existing method to parse for path
```
this sounds like a trick that will be hard to remember and maintain. A specific method should be extracted

bfitzpat replied:
It's going to be difficult to avoid doing some kind of trick to continue to use the URL query parsing. Otherwise we will need to write our own parsing algorithm, which I think is more error prone. I am open to suggestions on how to do this in a manageable fashion.

apupier replied:
tried to have a look. I do not understand why there is a need for the didact suffix.
If I remove the if block, the tests are still passing.

apupier replied:
abandoning to refactor it. it is effectively complicated code to understand as it seems to sometimes use didact prefix, sometimes not.
it sound so tnot be the same path used in extensionFunction.test.ts and commandHandler.test.ts which is used.

bfitzpat replied:
It took me a bit to process what you were saying here, but eventually I figured it out. Thanks. Have corrected the problems, added the additional test, and updated the example.

apupier commented on src/extensionFunctions.ts:
```diff
@@ -896,4 +899,41 @@ export function sendTextToNamedOutputChannel(message: string, channelName?: stri
+	} else {
+		let content : string | undefined = undefined;
+		if (out.scheme === 'file') {
+			await getDataFromFile(out)
+				.then(async fileContent => {
+					content = fileContent;
+				}).catch( (error) => {
+					showFileUnavailable(error);
+				});
```
I think it is better to avoid mixing `await` and `then`:
```
try {
content= await getDataFromFile(out);
} catch(error) {
showFileUnavailable(error);
}
```

apupier commented on src/extensionFunctions.ts:
```diff
@@ -896,4 +899,41 @@ export function sendTextToNamedOutputChannel(message: string, channelName?: stri
+					showFileUnavailable(error);
+				});
+		} else if (out.scheme === 'http' || out.scheme === 'https'){
+			const urlToFetch = out.toString();
+			await getDataFromUrl(urlToFetch)
+				.then(async fileContent => {
+					content = fileContent;
+				}).catch( (error) => {
+					showFileUnavailable(error);
+				});
```
I think it is better to avoid mixing `await` and `then`

apupier commented on src/test/suite/extensionFunctions.test.ts:
```diff
@@ -283,6 +283,39 @@ suite('Extension Functions Test Suite', () => {
+		expect(textInClipBoard).to.be.equal('a test with a %24 percentage inside');
+	});
+
+	test('test copy to clipboard with reported failed string', async() => {
```
I think that the naming of the text is not giving enough information in the context of the code.

apupier commented on src/test/suite/extensionFunctions.test.ts:
```diff
@@ -283,6 +283,39 @@ suite('Extension Functions Test Suite', () => {
+		expect(clipboard_content2).to.equal("The fox jumped over the lazy dog again.");
+	});
+
+	test('test copy to clipboard', async() => {
+		await extensionFunctions.placeTextOnClipboard('a test');
+		const textInClipBoard: string = await vscode.env.clipboard.readText();
+		expect(textInClipBoard).to.be.equal('a test');
+	});
```
this test is redundant with `test copy text to clipboard command`

apupier commented on src/test/suite/extensionFunctions.test.ts:
```diff
@@ -283,6 +283,39 @@ suite('Extension Functions Test Suite', () => {
+		const textForClipboard = 'The fox jumped over the lazy dog.';
+		await extensionFunctions.placeTextOnClipboard(textForClipboard);
+		const clipboard_content = await vscode.env.clipboard.readText();
+		expect(clipboard_content).to.equal("The fox jumped over the lazy dog.");
```
variable `textForClipboard` can be used

apupier commented on src/test/suite/extensionFunctions.test.ts:
```diff
@@ -283,6 +283,39 @@ suite('Extension Functions Test Suite', () => {
+	test('test copy text to clipboard command', async function() {
+		const textForClipboard = 'The fox jumped over the lazy dog.';
+		await extensionFunctions.placeTextOnClipboard(textForClipboard);
+		const clipboard_content = await vscode.env.clipboard.readText();
```
camel case is used everywhere in our codebase. maybe better to stick with this notation

apupier commented on src/test/suite/commandHandler.test.ts:
```diff
@@ -101,6 +101,13 @@ suite("Command Handler tests", function () {
 		await resetAfterTest();
 	});
+
+	test.skip('Reproducer FUSETOOLS-706 - $$ in URI', async() => {
+		const didactUri = 'didact://?commandId=vscode.didact.copyToClipboardCommand&text=%5BSend%20some%20fantastic%20text%20to%20a%20Terminal%20window%21%5D%28didact%3A%2F%2F%3FcommandId%3Dvscode.didact.sendNamedTerminalAString%26text%3DTerminalName%24%24echo%2BDidact%2Bis%2Bfantastic%2521%29';
+		await processInputs(didactUri);
+		const textInClipBoard: string = await vscode.env.clipboard.readText();
+		expect(textInClipBoard).to.be.equal('[Send some fantastic text to a Terminal window!](didact://?commandId=vscode.didact.sendNamedTerminalAString&text=TerminalName$$echo+Didact+is+fantastic%21)');
+	});	
```
this test is not related to the addition of the `copyFileTextToClipboardCommand`

bfitzpat replied:
Correct, but it is related to the initial difficulty with the doubly URL encoded text, which is related.

apupier commented on src/test/suite/extensionFunctions.test.ts:
```diff
@@ -283,6 +283,39 @@ suite('Extension Functions Test Suite', () => {
+
+	test('test copy file text to clipboard command', async function() {
+		const filePathForClipboard = vscode.Uri.parse('didact://?commandId=vscode.didact.copyFileTextToClipboardCommand&extension=src/test/data/textForClipboard.txt');
+		await vscode.env.clipboard.writeText('');
```
it would be better to use an afterEach that is setting back the clipboard.

apupier commented on src/extensionFunctions.ts:
```diff
@@ -896,4 +899,41 @@ export function sendTextToNamedOutputChannel(message: string, channelName?: stri
+	} else {
+		let content : string | undefined = undefined;
+		if (out.scheme === 'file') {
+			await getDataFromFile(out)
```
what about using directly` fs.readFileSync(uri.fsPath, 'utf8');`?
it will:
- avoid to reuse the getDatafromFile which is doing a transformation for Didact files and not for this case. (leading to think that the getDataFormFile might be better named to have the information that it is providing the transformed value of adoc/md files but better in another PR)
- allows to copy raw data from a `adoc` or `md` file
- avoid extra filename check

apupier reviewed (changes requested):
adding the new method copyFiletextToClipboard is different than the unsupported $$. I think that 2 different JIRAs should be created to make it clearer.

bfitzpat commented:
> adding the new method copyFiletextToClipboard is different than the unsupported $$. I think that 2 different JIRAs should be created to make it clearer.

I'm going to disagree, since the $$ issue led to development of the new command.

apupier commented:
> > adding the new method copyFiletextToClipboard is different than the unsupported $$. I think that 2 different JIRAs should be created to make it clearer.
> 
> I'm going to disagree, since the $$ issue led to development of the new command.

it is the combination of the wrong encoding of the URL and the unsupported $$ in parameters for text and user.

Even if you think that it is only the $$ that led to the development of this new feature, it stays clearer to have 2 different issues.

bfitzpat commented:
@apupier after I parsed your comments/responses, I finally figured out what you were after here and discovered the problem you mentioned. I have corrected the issues, removed the prefix, added the additional test, and updated the example. I think the test between the command handler and the extension functions tests are very similar, but processed differently enough that the overlap is still valid.

bfitzpat commented on src/test/suite/extensionFunctions.test.ts:
```diff
@@ -283,6 +287,33 @@ suite('Extension Functions Test Suite', () => {
+		expect(textInClipBoard).to.be.equal(percentTextForClipboard);
+	});
+
+	test('test copy to clipboard with doubly url encoded text for long didact link', async() => {
```
I think it's fine the way it is. But thanks

lhein commented on src/extensionFunctions.ts:
```diff
@@ -896,4 +905,35 @@ export function sendTextToNamedOutputChannel(message: string, channelName?: stri
+		let content : string | undefined = undefined;
+		if (out.scheme === 'file') {
+			try {
+				content = fs.readFileSync(out.fsPath, 'utf8');
```
There is also the VSCode internal implementation of the fs library which is not blocking. Maybe you can take a look if it is possible to use it instead of the external one?

bfitzpat replied:
I have looked at that in the past and so far I have only found the internal library useful when simply checking to see if we can get file contents, but not in actually retrieving the file contents. For instance https://github.com/redhat-developer/vscode-didact/blob/e03cf0bd48217b5c24134de7ec2e354cf3d72be7/src/extensionFunctions.ts#L735 just uses the internal implementation to make sure we can get something out of the file, but we're using the fs.readFileSync call most places to get the actual data out as a UTF-8 encoded string. 

I have not found another useful way to play with that internal fs version yet.

bfitzpat replied:
@lhein provided an example

```
const f = await vscode.workspace.fs.readFile(vscode.Uri.file(file));
const yamlDoc:DataVirtConfig = YAML.parse(f.toString());
```

I will spin this off into a separate JIRA to explore, since we use the external FS in many places and it might be useful to do a migration.

lhein commented on src/extensionFunctions.ts:
```diff
@@ -896,4 +905,35 @@ export function sendTextToNamedOutputChannel(message: string, channelName?: stri
+			} catch(error) {
+				showFileUnavailable(error);
+			}
+		}
```
We are lacking an `else` here which actually is not a big issue but whenever someone tries with a protocol other than file or http(s) then there will be nothing in the logs that could help finding the problem from the users point of view. I would suggest to at least put out a message to the output channel or log to signal an unsupported scheme/protocol.

bfitzpat replied:
Added the else case
